File: backend/app/execution/capabilities/fill_series.py
"""fillSeries — generate a number / date / weekday / pattern series into a range."""

from __future__ import annotations

from datetime import date, timedelta
from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range
from app.execution.utils import parse_date_flexible


def _add_unit(d: date, step: int, unit: str) -> date:
    if unit == "day":
        return d + timedelta(days=step)
    if unit == "week":
        return d + timedelta(weeks=step)
    if unit == "month":
        mo = d.month - 1 + step
        y = d.year + mo // 12
        mo = mo % 12 + 1
        # Clamp day to end-of-month to avoid Feb-31.
        try:
            return date(y, mo, d.day)
        except ValueError:
            return date(y, mo, 28)
    if unit == "year":
        try:
            return date(d.year + step, d.month, d.day)
        except ValueError:
            return date(d.year + step, d.month, 28)
    return d


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("range")
    series_type = params.get("seriesType")
    start = params.get("start")
    step = float(params.get("step", 1))
    pattern = params.get("pattern")
    date_unit = params.get("dateUnit", "day")
    count = params.get("count")
    horizontal = bool(params.get("horizontal", False))

    if not address or not series_type:
        return {"status": "error", "message": "fillSeries requires 'range' and 'seriesType'."}

    rng = resolve_range(ctx.workbook_handle, address)
    total_cells = count if count else (rng.columns.count if horizontal else rng.rows.count)
    total_cells = int(total_cells or 0)
    if total_cells <= 0:
        return {"status": "success", "message": "No cells to fill."}

    series: list[Any] = []
    if series_type == "number":
        s = float(start) if start is not None else 1.0
        for i in range(total_cells):
            series.append(s + i * step)
    elif series_type == "date":
        d = parse_date_flexible(start) or date.today()
        for i in range(total_cells):
            series.append(_add_unit(d, int(i * step), date_unit).strftime("%d/%m/%Y"))
    elif series_type == "weekday":
        d = parse_date_flexible(start) or date.today()
        while d.weekday() >= 5:  # Sat=5, Sun=6
            d += timedelta(days=1)
        for _ in range(total_cells):
            series.append(d.strftime("%d/%m/%Y"))
            d += timedelta(days=int(step))
            while d.weekday() >= 5:
                d += timedelta(days=1)
    elif series_type == "repeatPattern":
        if not pattern:
            return {"status": "error", "message": "seriesType='repeatPattern' requires 'pattern'."}
        for i in range(total_cells):
            series.append(pattern[i % len(pattern)])
    else:
        return {"status": "error", "message": f"Unknown seriesType: {series_type}"}

    if ctx.dry_run:
        return {"status": "preview", "message": f"Would fill {total_cells} {series_type} value(s) to {rng.address}."}

    grid = [series] if horizontal else [[v] for v in series]
    sheet = rng.sheet
    top_left = rng[0, 0]
    try:
        target = sheet.range(
            (top_left.row, top_left.column),
            (top_left.row + len(grid) - 1, top_left.column + len(grid[0]) - 1),
        )
        target.value = grid
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Write failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Wrote {total_cells} {series_type} value(s) to {rng.address}.",
        "outputs": {"range": rng.address, "filledCount": total_cells},
    }


registry.register("fillSeries", handler, mutates=True)

File: frontend/src/engine/capabilities/fillSeries.ts
/**
 * fillSeries — generate a series (number / date / weekday / repeat pattern).
 *
 * Replaces the common pattern where the LLM builds a manual 2D array of
 * sequential numbers or dates in writeValues. A dedicated primitive is safer
 * for long sequences and more discoverable.
 *
 * Single context.sync: build the series in JS, write back.
 */

import { CapabilityMeta, FillSeriesParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";
import { parseDateFlexible, formatDateDMY } from "../utils/parseDateFlexible";

const meta: CapabilityMeta = {
  action: "fillSeries",
  description: "Generate a number, date, weekday, or repeating-pattern series into a range",
  mutates: true,
  affectsFormatting: false,
};

function addDateStep(d: Date, step: number, unit: "day" | "week" | "month" | "year"): Date {
  const out = new Date(d);
  switch (unit) {
    case "day":   out.setUTCDate(out.getUTCDate() + step); break;
    case "week":  out.setUTCDate(out.getUTCDate() + step * 7); break;
    case "month": out.setUTCMonth(out.getUTCMonth() + step); break;
    case "year":  out.setUTCFullYear(out.getUTCFullYear() + step); break;
  }
  return out;
}

async function handler(
  context: Excel.RequestContext,
  params: FillSeriesParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const {
    range: address,
    seriesType,
    start = seriesType === "number" ? 1 : undefined,
    step = 1,
    pattern,
    dateUnit = "day",
    count,
    horizontal = false,
  } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would fillSeries ${seriesType} into ${address}` };
  }

  options.onProgress?.(`Generating ${seriesType} series...`);

  const range = resolveRange(context, address);
  range.load(["rowCount", "columnCount", "address"]);
  await context.sync();

  // How many cells we need to fill.
  const totalCells = (count !== undefined)
    ? count
    : horizontal
      ? range.columnCount
      : range.rowCount;

  const series: (string | number)[] = [];

  switch (seriesType) {
    case "number": {
      const s = typeof start === "number" ? start : Number(start ?? 1);
      for (let i = 0; i < totalCells; i++) series.push(s + i * step);
      break;
    }
    case "date": {
      const s = parseDateFlexible(start ?? new Date()) ?? new Date();
      for (let i = 0; i < totalCells; i++) {
        series.push(formatDateDMY(addDateStep(s, i * step, dateUnit)));
      }
      break;
    }
    case "weekday": {
      let d = parseDateFlexible(start ?? new Date()) ?? new Date();
      // Advance to first weekday if start lands on a weekend.
      while (d.getUTCDay() === 0 || d.getUTCDay() === 6) d = addDateStep(d, 1, "day");
      for (let i = 0; i < totalCells; i++) {
        series.push(formatDateDMY(d));
        do { d = addDateStep(d, step, "day"); } while (d.getUTCDay() === 0 || d.getUTCDay() === 6);
      }
      break;
    }
    case "repeatPattern": {
      if (!pattern || pattern.length === 0) {
        return { stepId: "", status: "error", message: "pattern is required for seriesType='repeatPattern'." };
      }
      for (let i = 0; i < totalCells; i++) {
        const v = pattern[i % pattern.length];
        series.push(typeof v === "boolean" ? (v ? "TRUE" : "FALSE") : (v as string | number));
      }
      break;
    }
  }

  // Shape into 2D for write.
  const grid: (string | number)[][] = horizontal
    ? [series]
    : series.map((v) => [v]);

  // Fit to the provided range dimensions, resizing if needed.
  const topLeft = range.getCell(0, 0);
  const rows = grid.length;
  const cols = grid[0]?.length ?? 1;
  const target = topLeft.getResizedRange(rows - 1, cols - 1);
  try {
    target.values = grid;
    await context.sync();
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    return { stepId: "", status: "error", message: `Failed to write series: ${msg}`, error: msg };
  }

  return {
    stepId: "",
    status: "success",
    message: `Wrote ${totalCells} ${seriesType} value(s) to ${address}.`,
    outputs: { range: address, filledCount: totalCells },
  };
}

registry.register(meta, handler as any);
export { meta };
